cmd/registry-replacer/main.go

```diff
@@ -0,0 +1,221 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+)
+
+func main() {
+	configDir := flag.String("config-dir", "", "The directory with the ci-operator configs")
+	flag.Parse()
+	if *configDir == "" {
+		logrus.Fatal("--config-dir arg is required")
+	}
+
+	var errs []error
+	errLock := &sync.Mutex{}
+	wg := sync.WaitGroup{}
+	if err := config.OperateOnCIOperatorConfigDir(
+		*configDir,
+		func(config *api.ReleaseBuildConfiguration, info *config.Info) error {
+			wg.Add(1)
+			go func(filename string) {
+				defer wg.Done()
+				if err := replacer(
+					githubFileGetterFactory,
+					func(data []byte) error {
+						return ioutil.WriteFile(filename, data, 0644)
+					})(config, info); err != nil {
+					errLock.Lock()
+					errs = append(errs, err)
+					errLock.Unlock()
+				}
+			}(info.Filename)
+			return nil
+		},
+	); err != nil {
+		logrus.WithError(err).Fatal("Failed to operate on ci-operator-config")
+	}
+	wg.Wait()
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		logrus.WithError(err).Fatal("Encountered errors")
+	}
+}
+
+// replacer ensures replace directives are in place. It fetches the files via http because using git
+// en masse easily kills a developer laptop whereas the http calls are cheap and can be parallelized without
+// bounds.
+func replacer(
+	githubFileGetterFactory func(org, repo, branch string) githubFileGetter,
+	writer func([]byte) error,
+) func(*api.ReleaseBuildConfiguration, *config.Info) error {
+	return func(config *api.ReleaseBuildConfiguration, info *config.Info) error {
+		if len(config.Images) == 0 {
+			return nil
+		}
+
+		originalConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal config for comparison: %w", err)
+		}
+
+		for idx := range config.Images {
+			foundTags, err := ensureReplacement(&config.Images[idx], githubFileGetterFactory(info.Org, info.Repo, info.Branch))
+			if err != nil {
+				return fmt.Errorf("failed to ensure replacements: %w", err)
+			}
+			for _, foundTag := range foundTags {
+				if config.BaseImages == nil {
+					config.BaseImages = map[string]api.ImageStreamTagReference{}
+				}
+				config.BaseImages[foundTag.String()] = api.ImageStreamTagReference{
```

Comment:
Very unlikely but we should also guard against changing this if it's already set somehow. I think we will want the "overwrite existing config to do the right thing" to be an opt-in flag that we will turn on once we alert everyone we're doing it, but before that happens we should make sure we're not ever making a change that could change user's config outside of what we're already doing.

Reply:
Updated